Fixed support for join entities

jebw opened the pull request:
### Jira link

HOTT-????

### What?

I have added/removed/altered:

- [x] Added support for parsing join entities (ie entities which only provide relationships between other entities but no attributes of their own
- [x] Extended the jsonapi parser specs
- [x] Added error handling when receiving unexpected jsonapi format content

### Why?

I am doing this because:

- The permutations additions to the commodities endpoint cannot currently be parsed by this gem, preventing enabling of that feature

willfish commented:
How do you feel about using this jsonapi parser in the frontend?

willfish commented:
I prefer your explicit approach to validating different fixtures and I think we keep making tweaks to support different facets of the jsonapi protocol and end up needing to duplicate things

jebw commented:
To use the parser in the frontend, we need several things I think

1. Split out the parser itself from the data models (separate gem, which uktt depends upon?) - otherwise we end up pulling the DC models into the Frontend which is more to unpick before we can use it.
2. This parser is actually behind the one in the frontend and several of the fixes we've introduced to the frontend over the last few months would end up getting dropped, so the specs need enhancing and the fixes porting over
3. Make uktt, frontend and admin depend upon the new `uktt-jsonapi-parser` gem
4. Ideally port the faraday jsonapi middleware layer from the admin to the new gem and drop the custom code from DC and Frontend
